persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java

```diff
@@ -187,271 +190,268 @@ protected RocksDB openDatabase(Path location, Options options) throws RocksDBExc
     }
 
     @Override
-    public void stop() {
-        try {
-            semaphore.acquire(Integer.MAX_VALUE);
-        } catch (InterruptedException e) {
-            throw new PersistenceException("Cannot acquire semaphore", e);
-        }
-        try {
+    public CompletionStage<Void> stop() {
+        return blockingManager.runBlocking(() -> {
             handler.close();
             expiredDb.close();
-        } finally {
-            stopped = true;
-            semaphore.release(Integer.MAX_VALUE);
-        }
-    }
-
-    @Override
-    public void destroy() {
-        stop();
-        Util.recursiveFileRemove(getLocation().toFile());
-        Util.recursiveFileRemove(getExpirationLocation().toFile());
+        }, "rocksdb-stop");
     }
 
     @Override
-    public boolean isAvailable() {
-        return getLocation().toFile().exists() && getExpirationLocation().toFile().exists();
+    public Set<Characteristic> characteristics() {
+        return EnumSet.of(Characteristic.BULK_READ, Characteristic.EXPIRATION, Characteristic.SEGMENTABLE);
     }
 
     @Override
-    public void clear() {
-        handler.clear(null);
+    public CompletionStage<Boolean> isAvailable() {
+        return blockingManager.supplyBlocking(() -> getLocation().toFile().exists() && getExpirationLocation().toFile().exists(),
+                "rocksdb-available");
     }
 
     @Override
-    public void clear(IntSet segments) {
-        handler.clear(segments);
+    public CompletionStage<Void> clear() {
+        return handler.clear();
     }
 
     @Override
-    public int size() {
-        return handler.size(null);
-    }
-
-    @Override
-    public int size(IntSet segments) {
+    public CompletionStage<Long> size(IntSet segments) {
         return handler.size(segments);
     }
 
     @Override
-    public boolean contains(Object key) {
-        return handler.contains(-1, key);
+    public CompletionStage<Long> approximateSize(IntSet segments) {
+        return handler.approximateSize(segments);
     }
 
     @Override
-    public boolean contains(int segment, Object key) {
-        return handler.contains(segment, key);
-    }
-
-    @Override
-    public Publisher<K> publishKeys(Predicate<? super K> filter) {
-        return handler.publishKeys(null, filter);
+    public CompletionStage<Boolean> containsKey(int segment, Object key) {
+        // This might be able to use RocksDB#keyMayExist - but API is a bit flaky
+        return load(segment, key)
```

Comment:
Performing a full load seems wasteful here, as it means that the value has to be unmarshalled even though it is never used.

Reply:
This is the same as it was prior, but I can see if I can add it easily enough.